lib/application/widgets/theme/custom_button.dart

```diff
@@ -0,0 +1,313 @@
+class CustomButton extends StatefulWidget {
+  const CustomButton({
+    required this.text,
+    this.onPressed,
+    this.isPressedOverlayEnabled = false,
+    this.leadingWidgetBuilder,
+    this.borderBuilder,
+    this.backgroundColorBuilder,
+    this.backgroundGradientBuilder,
+    this.textStyleBuilder,
+  });
+
+  /// The button text.
+  ///
+  /// Its style may be customized by implementing [textStyleBuilder].
+  final String text;
+
+  final VoidCallback? onPressed;
+
+  /// If `true` adds a overlay over the button when it's pressed.
+  final bool isPressedOverlayEnabled;
+
+  /// Builds the button leading widget based on the button `state`.
+  final Widget Function(ButtonState state)? leadingWidgetBuilder;
+
+  /// Builds the button border based on the button `state`.
+  ///
+  /// If `null` no border is added.
+  final BoxBorder? Function(ButtonState state)? borderBuilder;
+
+  /// Builds the button background color based on the button `state`.
+  ///
+  /// Will be overriden by [backgroundGradientBuilder] if implemented.
+  /// If `null` no background color is added.
+  final Color? Function(ButtonState state)? backgroundColorBuilder;
+
+  /// Builds the button background gradient based on the button `state`.
+  ///
+  /// Overrides [backgroundColorBuilder] if both are implemented.
+  /// If `null` no background gradient is added.
+  final LinearGradient? Function(ButtonState state)? backgroundGradientBuilder;
+
+  /// Builds the button [text] style based on the button `state`.
+  ///
+  /// If `null` the default text theme is used.
+  final TextStyle? Function(ButtonState state)? textStyleBuilder;
+
+  @override
+  CustomButtonState createState() => CustomButtonState();
+}
+
+@visibleForTesting
+class CustomButtonState extends State<CustomButton> {
+  late ButtonState state;
+
+  @override
+  void initState() {
+    state = widget.onPressed != null ? ButtonState.normal : ButtonState.disabled;
+    super.initState();
+  }
+
+  @override
+  void didUpdateWidget(covariant CustomButton oldWidget) {
+    if (widget.onPressed != oldWidget.onPressed) {
+      setState(() {
+        state = widget.onPressed != null ? ButtonState.normal : ButtonState.disabled;
+      });
+    }
+    super.didUpdateWidget(oldWidget);
+  }
+
+  @override
+  Widget build(BuildContext context) {
+    final rowWidgets = <Widget>[];
+
+    if (widget.leadingWidgetBuilder != null) {
+      rowWidgets.addAll([
+        widget.leadingWidgetBuilder!(state),
+        context.horizontalBox(Spacing.xSmall),
+      ]);
+    }
+
+    rowWidgets.add(Text(widget.text, style: widget.textStyleBuilder?.call(state)));
+
+    final buttonContent = InkWell(
+      // Disables default splash/highlight animation
+      splashColor: Colors.transparent,
+      highlightColor: Colors.transparent,
+      onTap: widget.onPressed,
+      onHighlightChanged: (isPressed) {
+        setState(() {
+          state = isPressed ? ButtonState.pressed : ButtonState.normal;
+        });
```

Comment:
We should use hooks here

Reply:
We are using `StatefulWidget` because we need to update the button's state based on `void didUpdateWidget(covariant CustomButton oldWidget)` event. I'm not sure if we can achieve such behavior with hooks.

Reply:
Hooks can achieve anything a stateful widget can, just look at any example in this repository or the `flutter_hooks` itself.